### ISSUE TITLE
Access current hostname / port from running server

### ISSUE THREAD

**Issue by BlameOmar** [vapor/vapor#2012]
I'm a big fan of testing, and one of my favorite ways to test servers is to spin one up in the test process and start sending it requests programmatically. To that end, it's useful to configure the server to bind to an arbitrary open port so that the success of the tests is not dependent on the availability of a statically configured port. This is especially important if the tests in the test suite are being executed in parallel (in which case, there may be multiple servers) or if the test suite itself is being executed on multiple versions of the code at once.

Currently Vapor does not support binding to an arbitrary port in a useful manner, even though it's possible to do.

### Steps to reproduce

* Specify `0` as the serving port and start the server. In addition to simply supplying `--port 0` as command line arguments, there are a few ways to do this in code:

```swift
// test.swift
app = try! App.app(for: .testing)
try! app.make(Server.self).start(hostname: nil, port: 0).wait()
```

```swift
// configure.swift
// getAppServices(for environment: Environment) throws -> Services
// …
var services = Services.default()
var httpConfig = NIOServerConfig.default()
httpConfig.port = 0
services.register(httpConfig)
// …

// test.swift
// class func setUp()
// …
app = try! App.app(for: .testing)  // Calls getAppServices(for:)
try! app.make(Server.self).start(hostname: nil, port: nil).wait()
// …
```

### Expected behavior

* The server starts on an arbitrary port.
* The serving port is printed to the console.
* API exists to get the serving port.

### Actual behavior

* The server starts on an arbitrary port (this can be confirmed using the various platform tools available to get the open ports of a process; I checked `Activity Monitor.app`).
* Console: `Server starting on http://localhost:0`
* No API exists to get the serving port.

### Environment
* Vapor Framework version: 3.3.0
* Vapor Toolbox version: 3.1.10
* OS version: macOS 10.14.5

### Further details/thoughts

Regarding the lack of API, my first thought was to use `NIOServerConfig.port`.
```swift
try! app.make(NIOServerConfig.self).port
```
However, that statement does not return the runtime port value. Depending on whether the port was specified in the start method (as in the first example), or the port was specified in the server configuration (as in the latter example), the value is either the default of `8080` or `0`. A quick search of the repository for "port" doesn't yield any other APIs that might be useful. In fact, it seems that the package's own [tests] simply hardcode port(s) that are assumed to be available!

[tests]: https://github.com/vapor/vapor/blob/master/Tests/VaporTests/ApplicationTests.swift

A simple solution would be to add a new protocol requirement to Server:

```swift
var port: Int? { get }
```

This port would be the runtime port of the server, which may be different from the configured value. The return value is `Int?`, same as the similar property on `SocketAddress`. A default implementation could be provided to return `nil`, which would avoid making this a breaking change to `Server`; otherwise, this is a breaking change and must be released in a major version. Regardless, `nil` is an appropriate value for a custom implementation of `Server` that is serving from a UNIX socket (although I'm not sure what it would do for `start(hostname: port:)`; throw perhaps?).

One limitation of the proposed API is that it further bakes in the current assumption that there is a single host/port, which may not hold for an arbitrary implementation of `Server`. Perhaps there should be a `SingleIPSocketServer` intermediary protocol that contains this property, along with `start(hostname: port:)`, with `Server` containing `onShutdown` and similar APIs.

It would also be nice to have a similar property for the host IP address(es), but that's out of scope for this bug.

More generally, it would be useful to have the additional concept of runtime configuration for Servers and other Services. It's not difficult to imagine having other parameters that are only known until runtime, especially if you have some fancy auto-tuning software that's doing the setting of them, and it would be nice if for every Service.Configuration type, it was possible to have some Runtime<Service.Configuration> type where all the "auto" values have been resolved to their current ones. Admittedly, this is probably a bit much to get into Vapor 4 unless there already have been discussions about it.

**Comment by BlameOmar** [comment]
I created a couple of PRs to add a read-only port property to HTTPServer.

**Comment by tanner0101** [comment]
+1 to adding this. vapor/http is no longer used in Vapor 4, but I've extracted the useful bits here:

```swift
guard let port_uint16 = channel.localAddress?.port else { return nil }
return Int(exactly: port_uint16)
```

**Comment by biboran** [comment]
It seems in Vapor 4 you can get the port like this:
```
(app.server as? HTTPServer).localAddress?.port
```

**Comment by 0xTim** [comment]
Yes this can be solved with `app.http.server.shared.localAddress`, e.g. https://github.com/vapor/vapor/blob/main/Tests/VaporTests/ApplicationTests.swift#L142-L174
